### PR TITLE
API: Do not pre-encode TEAL byte values

### PR DESCRIPTION
**Pull request by jasonpaulos**
## Summary

This PR properly types our REST API fields `TealValue.bytes` and `TealKeyValue.key` as binary data.

The result is that our generated types `model.TealValue` and `model.TealKeyValue` will use `[]byte` instead of `string` for these fields. This will let us store binary data there directly instead of having to base64 encode it manually. When JSON encoding, `[]byte` is automatically base64 encoded and decoded for us, so this should have no effect on consumers.

The real benefit to this change is that we can now encode the exact same `model.TealValue` struct to both JSON and msgpack properly, since both formats can encode and decode `[]byte`. If we wanted to do this without this change, we'd either need to change whether `model.TealValue.Bytes` is a base64 encoded string or not depending on the encoding method, or always leave it as a base64 encoded string and accept that we'd encode this verbatim in our msgpack responses.

I could not find any instances where we msgpack encode `model.TealValue` or `model.TealKeyValue` in our current code base (reviewers, please let me know I got this wrong, that would mean this PR *is* breaking), but we plan to do so in the future, specifically #5440

### Still to do
- [ ] Ensure downstream dependencies (SDKs) do not have a breaking change because of this

## Test Plan

- [x] Existing tests modified
- [ ] New unit test to ensure encoding works as expected

### CODE REVIEW

File: daemon/algod/api/algod.oas3.yml [hunk 1]
```diff
@@ -2174,6 +2174,8 @@
         "description": "Represents a key-value pair in an application store.",
         "properties": {
           "key": {
+            "format": "byte",
+            "pattern": "^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$",
```
Comment (jannotti) [inline daemon/algod/api/algod.oas3.yml:2178]: Is it impossible for b64 encodings to end in three equal signs?

Comment (jannotti) [reply daemon/algod/api/algod.oas3.yml:2178]: I guess not. Neat.

**Review by ahangsu — COMMENTED** [review]
looks mostly making sense to me, just trying to understand the comments:

> If we wanted to do this without this change, we'd either need to change whether `model.TealValue.Bytes` is a base64 encoded string or not depending on the encoding method, or always leave it as a base64 encoded string and accept that we'd encode this verbatim in our msgpack responses.

So if I understood correctly, json always convert `[]byte` into b64 encoding, while msgpack treat them as byte array, without any encoding conversion?

**Comment by jasonpaulos** [comment]
> So if I understood correctly, json always convert `[]byte` into b64 encoding, while msgpack treat them as byte array, without any encoding conversion?

Yes that's correct. Msgpack is a binary format and it's capable of directly encoding byte strings.

**Comment by jasonpaulos** [comment]
Closing because after discussion with @ahangsu, we decided the problems with the existing `TealValue` are too great to be fixed without breaking changes to our existing endpoints.

To summarize, the problems are:
1. `TealValue.bytes` and `TealKeyValue.key` are typed as strings, not binary data.
2. `TealValue` marks both `bytes` and `uint` as required, even though they are mutually exclusive.

This PR attempts to address the first problem, and while it does succeed here, even if merged we'd want to add a special case to https://github.com/algorand/generator so that our SDKs do not introduce breaking changes.

The second problem cannot be addressed without a breaking change.

For these reasons #5440 will introduce a new type, and `TealValue` should be considered deprecated

**Pull request closed without merging**